clinto/parsers/base.py

```diff
@@ -0,0 +1,63 @@
+from __future__ import absolute_import
+import json
+import os
+import copy
+
+
+def update_dict_copy(a, b):
+    temp = copy.deepcopy(a)
+    temp.update(b)
+    return temp
+
+
+class ClintoArgumentParserException(Exception):
+    def __init__(self, parser, *args, **kwargs):
+        self.parser = parser
+
+
+def parse_args_monkeypatch(self, *args, **kwargs):
+    raise ClintoArgumentParserException(self)
+
+
+class BaseParser(object):
+
+    extensions = []
+    contains = ""
+
+    def __init__(self, script_path=None, script_source=None):
+        self.is_valid = False
+        self.error = ''
+        self.parser = None
+
+        self.script_path = script_path
+        self.script_source = script_source
+
+        if not self.check_valid():
+            return
+
+        self.extract_parser()
+
+        if self.parser is None:
+            return
+
+        self.process_parser()
+
+    def check_valid(self):
+        if self.script_path:
+            ext = os.path.splitext(os.path.basename(self.script_path))[1]
+            return ext in self.extensions and self.contains in self.script_source
```

Comment:
The import is coming from the package, it's a module. from . import CustomParser should be from mymodule.templates import CustomParser.

I did it by loading the file and then making diagnostics, like this:

+SUPPORTED_PARSERS = {
+    'argparse': {
+        'methods': ['__package__', 'ArgumentParser'],
+        'parser': argparse_parser.ArgparseParser,
+    },
+    'click': {
+        'methods': ['__package__', 'parser', 'ClickException'],
+        'parser': click_parser#.ClickParser,
+    },
+}

That would go down the list of objects founds, and try to determine which parser is active via classes specific to that parser. That was the heuristic I was using. Though that obviously breaks sometimes from people putting parse_args where it shouldn't be.

Ultimately, I think we're going to have to experiment with AST to really get this right in all cases...